packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -509,13 +514,25 @@ function performRecordIdentifierUpdate(
 function detectMerge(
   keyOptions: KeyOptions,
   identifier: StableRecordIdentifier,
-  newId: string | null
+  data: ResourceIdentifierObject | ExistingResourceObject,
+  newId: string | null,
+  lids: IdentifierMap
 ): StableRecordIdentifier | false {
-  const { id } = identifier;
+  const { id, type, lid } = identifier;
   if (id !== null && id !== newId && newId !== null) {
     const existingIdentifier = keyOptions.id[newId];
 
     return existingIdentifier !== undefined ? existingIdentifier : false;
+  } else {
+    // detect unconfigured single-table polymorphism
+    // TODO deprecate in favor of users configuring lid generation
```

Comment:
I think the discussion we had was that we are not gonna require people to configure lid generation but rather find  way to get the request context to the identifier merge function so that we can do the merge every time. Can you please update the todo and file an issue?

Reply:
https://github.com/emberjs/data/issues/6890